Adds charts showing active Enrollments and Problem Difficulty

pomegranited opened the pull request:
Adds two more charts to the Open edX Dashboard:

![dashboard](https://user-images.githubusercontent.com/7556571/225819255-ea5f2eeb-ff7f-46fb-ab93-9f2ee01c8604.png)

**Course Enrollments**

![enrollments_pie](https://user-images.githubusercontent.com/7556571/225819273-a6f7316c-29ae-40b0-9dfd-b871fce75191.png)

* Pie chart
* Adds a dataset that links the MySQL course enrollments data with the course overviews, so we can show course titles.
* Protected by another row-level permission filter that targets this dataset.

**Problem Difficulty**

![problem_difficulty](https://user-images.githubusercontent.com/7556571/225819325-0cfded6b-0196-4389-84d8-156bee19b35d.png)

* Bar chart (untimed)
* Shows min, max, and average attempts people have made recorded against problem blocks.
* Adds a dataset that links the Clickhouse parsed xapi events to the Course Graph data, so we can see block display names.

Relates to https://github.com/openedx/openedx-oars/issues/30

**Testing instructions**

1. Checkout this branch.
2. If you have previously run `tutor config save` against an earlier version of this repo, you may have some old asset files hanging around. To erase them, run:
   ```
   rm -rf $(tutor config printroot)/env/plugins/oars/apps/data/assets/ 
   tutor config save
   ```
3. Re-init against your OARS tutor dev stack:
   ```
   tutor dev do init -l oars

   # load demo course and optional xapi data
   tutor dev do importdemocourse
   tutor dev do  load-xapi-test-data

   # and whatever users you want/need
   tutor dev do createuser --staff --superuser super_user super_user@email.com
   ```
1. Login to the LMS, enroll in the DemoX course, and complete some problems incorrectly and then correctly.
1. Login to Superset as either a DemoX course staff member, a global staff member, or a superuser
4. View the Dashboard, and ooh and ahh.
 
**Author Notes & Concerns**

a. Currently the Problem Difficulty chart shows attempt counts from all problem submissions, not just the ones that were eventually correct. Should I filter on correctness too (it's easy)?

openedx-webhooks commented:
Thanks for the pull request, @pomegranited!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

bmtcril reviewed (commented):
Enrollments is looking great, I ran into a couple of issues that I noted here just in case, but I know it's WIP

pomegranited commented on tutoroars/templates/oars/apps/pythonpath/superset-dashboard.py:
```diff
@@ -52,15 +73,25 @@ def update_assets():
+    timestamp = datetime.datetime.utcnow().replace(tzinfo=datetime.timezone.utc).isoformat()
     with zipfile.ZipFile(fp, "w") as ziph:  # ziph is zipfile handle
+        # Write a metadata.yaml file to the root dir
+        ziph.writestr(os.path.join(archive_base_dir, 'metadata.yaml'),
```
@bmtcril Ah.. I think that's because there's old OARS assets files still in your tutor config dirs. I've removed those `metadata.yaml` files here and am creating them on demand instead, so that I can import (and overwrite) each asset type individually. (I learned that Superset is fussy about both the `metadata.yaml` file contents and the base dir of the zip file when importing individual assets.)

Have updated the PR test instructions here.. could you try that with my latest code?

pomegranited commented on tutoroars/templates/oars/jobs/init/superset-init-security.sh:
```diff
@@ -27,69 +27,78 @@ role_name = "{{SUPERSET_OPENEDX_ROLE_NAME}}"
+    ),
+    (
+        "{{OPENEDX_MYSQL_DATABASE}}",
+        "Course Enrollments + Overview",
```
I think I've fixed this too, but will see what your testing shows.

pomegranited commented:
@bmtcril Ok, I think this is working now.. ready for your review!

bmtcril reviewed (approved):
I'm going to thumb this in spite of the error I saw, since there seems to be a workaround. I'm curious to hear if you see it too, though.

pomegranited commented:
@bmtcril 

>Here's as far as I've gotten (all working in tutor local):

Cool.. glad there's a workaround, and so hopefully that can lead to a fix!

> At which point everything ran (and said it was successful), but there was an error:
> `AssertionError: openedx.None table doesn't exist yet?`

Ah oops! I do get that error after rebuilding with my latest changes. It revealed two problems, which I fixed with 959fcdd.

* I renamed the dataset to remove the `+` sign from the name when I was trying to work out what was wrong with my asset imports. But I neglected to change it here. 
* The log message was using the null `{table}` variable instead of `{table_name}`, so wasn't helpful.

So now the Row Level Permissions Filter gets created as required, and now properly restricts access to course enrollments data.

Gonna merge this, and we can tweak things as needed.

openedx-webhooks commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/tutor-contrib-oars&entry.867055334=https://github.com/openedx/tutor-contrib-oars/pull/11&entry.1484655318=https://github.com/pomegranited&entry.752974735=2023-03-16+07:12&entry.1917517419=2023-03-20+17:56&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

pomegranited commented on tutoroars/templates/oars/apps/data/assets/datasets/OpenedX_Clickhouse/xAPI_Events_Course_Graph.yaml:
```diff
@@ -0,0 +1,90 @@
+  if(event.object.id like '%/xblock/%', splitByChar('/', event.object.id)[5], '')
+  as block_key, event.result.success as success, event.object.definition.extensions.'http://id.tincanapi.com/extension/attempt-id'
+  as attempts from xapi.xapi_events_all_parsed as xapi_events, coursegraph.coursegraph_nodes
+  as cg where cg.location = block_key
```
To make this dataset more broadly useful, it also needs to handle course type block_keys, and include all of the xapi data:

```
select cg.display_name, 
splitByChar('/', xapi_events.course_id)[-1] as course_key, 
if(event.object.id like '%/xblock/%', 
    splitByChar('/', event.object.id)[5], 
    format('block-v1:{0}+type@course+block@course', splitByChar(':', course_key)[2])
  ) as block_key, 
event.result.success as success,
event.object.definition.extensions.'http://id.tincanapi.com/extension/attempt-id' as attempts, 
xapi_events.*
FROM xapi.xapi_events_all_parsed as xapi_events, coursegraph.coursegraph_nodes as cg 
where block_key = cg.location
```

bmtcril replied:
Yeah I locally had a version of this which outer joined the coursegraph data so that everything would show up, but it's messy. I think that post-conference we should make a plan for the different datasets and document their quirks.

pomegranited replied:
Yes -- a plan would be great :)

I also think there's a better place for a lot of this detail, like a Clickhouse view.
